op-node/rollup/derive/engine_queue.go

```diff
@@ -224,7 +224,12 @@ func (eq *EngineQueue) Step(ctx context.Context) error {
 	}
 	outOfData := false
 	if len(eq.safeAttributes) == 0 {
-		eq.origin = eq.prev.Origin()
+		newOrigin := eq.prev.Origin()
+		// Check if the L2 safe head origin is consistent with the new origin
+		if err := eq.verifyNewL1Origin(ctx, newOrigin); err != nil {
```

Comment:
This seems odd that when we'll create a new safe head we verify local state against the unsafe head. It seems like it mostly works (two parts: 1st is I think we could get stuck in a spot where we have no safe attributes, but do have an invalid L2 chain, 2nd is accurately checking the canonical nature of the L1 origin).

In general, it seems like we should just be checking if the L1 origin of an unsafe block is canonical or not (& maybe we need to do this on a loop outside of inserting the unsafe blocks).

Reply:
I'm definitely keen to get @protolambda 's view on it, but from what I can see:
* if we have no safe attributes and an invalid L2 chain, this new verify method will trigger a reset and as part of doing that reset we'll go through `FindL2Heads` which filters the possible L2 unsafe heads and only considers those with a canonical L1 origin. So the reset process will ensure that we get back to having a valid L2 chain
* Checking just the unsafe head when the L1 origin changes makes sense to me because if the unsafe head is built on a canonical L1 origin then the whole L2 chain is too. If it's not we could potentially iterate back to find the latest suitable head but that's what reset is doing. And whether the L2 chain origin is canonical or not can only change when the L1 chain we're tracking changes (either by reorg or advancing) which is the L1 origin advancing. So I think this has the same result as checking each unsafe block but with fewer checks needed, at the cost of doing a full reset if any part of the unsafe chain is no longer built on canonical L1.

Reply:
Ok, I think I was actually confused about this.
If there is no next attributes, then we verify the L1 origin of the L2 *unsafe* head. This is a pretty good place to do this check. The other potential place is when accepting unsafe payloads, but this code path gets hit more often.